gwcs/wcs.py

```diff
@@ -332,7 +388,502 @@ def _invert(self, *args, **kwargs):
+        maxiter = kwargs.pop('maxiter', 40)
+        adaptive = kwargs.pop('adaptive', True)
+        detect_divergence = kwargs.pop('detect_divergence', True)
+        quiet = kwargs.pop('quiet', True)
+
+        args_shape = np.shape(args)
+        nargs = args_shape[0]
+        arg_dim = len(args_shape) - 1
+
+        if nargs != self.world_n_dim:
+            raise ValueError("Number of input coordinates is different from "
+                             "the number of defined world coordinates in the "
+                             f"WCS ({self.world_n_dim:d})")
+
+        if self.world_n_dim != self.pixel_n_dim:
+            raise NotImplementedError(
+                "Support for iterative inverse for transformations with "
+                "different number of inputs and outputs was not implemented."
+            )
+
+        # form equation:
+        def f(x):
+            return np.add(np.subtract(self.__call__(*x), argsi), x)
+
+        # initial guess:
+        if nargs == 2 and self._approx_inverse is None:
+            self._calc_approx_inv(max_inv_pix_error=5, inv_degree=None)
+
+        if self._approx_inverse is None:
+            if self.bounding_box is None:
+                x0 = np.ones(self.pixel_n_dim)
+            else:
+                x0 = np.mean(self.bounding_box, axis=-1)
+
+        if arg_dim == 0:
+            argsi = args
+
+            if nargs == 2 and self._approx_inverse is not None:
+                x0 = self._approx_inverse(*argsi)
+                if not np.all(np.isfinite(x0)):
+                    return [np.array(np.nan) for _ in range(nargs)]
+
+            try:
+                result = self._vectorized_fixed_point(
+                    x0, argsi, tolerance=tolerance,
+                    maxiter=maxiter, adaptive=adaptive,
+                    detect_divergence=detect_divergence,
+                    quiet=quiet
+                ).ravel().tolist()
+
+                return list(map(np.array, result))
+
+            except NoConvergence:
+                return [np.array(np.nan) for _ in range(nargs)]
+
+        else:
+            arg_shape = args_shape[1:]
+            nelem = np.prod(arg_shape)
+            inv = np.empty((nargs, nelem))
+
+            args = np.reshape(args, (nargs, nelem))
+
+            failed_result = np.full(nargs, np.nan)
+
+            x0 = np.array(self._approx_inverse(*args)).T
+
+            try:
+                result = self._vectorized_fixed_point(
+                    x0, args.T, tolerance=tolerance,
+                    maxiter=maxiter, adaptive=adaptive,
+                    detect_divergence=detect_divergence,
+                    quiet=quiet
+                ).ravel().tolist()
+
+                inv = result
+
+            except NoConvergence as e:
+                inv[:, e.divergent] = failed_result
+
+            return list(np.reshape(inv, args_shape))
+
+
+    def _vectorized_fixed_point(self, pix0, world, tolerance, maxiter, adaptive,
```

Comment:
This algorithm could use a better summary of what is going on.

Reply:
I'll copy/paste (+minor adjustment) comments from `astropy.wcs.WCS._all_world2pix` if you find that the function itself is OK and will be kept in the final version of this PR.